File: frontend-reshet/src/services/mcp.ts
import { httpClient } from "./http";

export type McpAuthMode = "none" | "static_bearer" | "static_headers" | "oauth_user_account";
export type McpApprovalPolicy = "ask" | "always_allow";

export interface McpServer {
  id: string;
  tenant_id: string;
  name: string;
  description?: string | null;
  server_url: string;
  transport: string;
  auth_mode: McpAuthMode;
  is_active: boolean;
  auth_config: Record<string, unknown>;
  auth_metadata: Record<string, unknown>;
  capability_snapshot: Record<string, unknown>;
  oauth_client_id?: string | null;
  oauth_client_registration: Record<string, unknown>;
  tool_snapshot_version: number;
  sync_status: string;
  sync_error?: string | null;
  last_tested_at?: string | null;
  last_synced_at?: string | null;
  created_at: string;
  updated_at: string;
  has_static_bearer_token: boolean;
  has_static_headers: boolean;
  has_oauth_client_secret: boolean;
}

export interface McpDiscoveredTool {
  id: string;
  server_id: string;
  snapshot_version: number;
  name: string;
  title?: string | null;
  description?: string | null;
  input_schema: Record<string, unknown>;
  annotations: Record<string, unknown>;
  tool_metadata: Record<string, unknown>;
  created_at: string;
}

export interface McpAccountConnection {
  server_id: string;
  user_id: string;
  status: string;
  scopes: unknown[];
  account_metadata: Record<string, unknown>;
  last_error?: string | null;
  access_token_expires_at?: string | null;
  refresh_token_expires_at?: string | null;
  last_refreshed_at?: string | null;
  last_used_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface McpAgentMount {
  id: string;
  agent_id: string;
  server_id: string;
  applied_snapshot_version: number;
  approval_policy: McpApprovalPolicy;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateMcpServerRequest {
  name: string;
  description?: string;
  server_url: string;
  auth_mode: McpAuthMode;
  static_bearer_token?: string;
  static_headers?: Record<string, string>;
  auth_config?: Record<string, unknown>;
  oauth_client_id?: string;
  oauth_client_secret?: string;
  is_active?: boolean;
}

export type UpdateMcpServerRequest = Partial<CreateMcpServerRequest>;

export const mcpService = {
  listServers() {
    return httpClient.get<McpServer[]>("/mcp/servers");
  },

  getServer(id: string) {
    return httpClient.get<McpServer>(`/mcp/servers/${id}`);
  },

  createServer(payload: CreateMcpServerRequest) {
    return httpClient.post<McpServer>("/mcp/servers", payload);
  },

  updateServer(id: string, payload: UpdateMcpServerRequest) {
    return httpClient.patch<McpServer>(`/mcp/servers/${id}`, payload);
  },

  testServer(id: string) {
    return httpClient.post<Record<string, unknown>>(`/mcp/servers/${id}/test`, {});
  },

  syncServer(id: string) {
    return httpClient.post<Record<string, unknown>>(`/mcp/servers/${id}/sync`, {});
  },

  listTools(id: string, snapshotVersion?: number) {
    const suffix = snapshotVersion ? `?snapshot_version=${snapshotVersion}` : "";
    return httpClient.get<McpDiscoveredTool[]>(`/mcp/servers/${id}/tools${suffix}`);
  },

  startAuth(id: string) {
    return httpClient.post<{ authorization_url: string }>(`/mcp/servers/${id}/auth/start`, {});
  },

  getMyConnection(id: string) {
    return httpClient.get<McpAccountConnection | null>(`/mcp/servers/${id}/account/me`);
  },

  disconnectMyConnection(id: string) {
    return httpClient.delete<{ ok: boolean }>(`/mcp/servers/${id}/account/me`);
  },

  listAgentMounts(agentId: string) {
    return httpClient.get<McpAgentMount[]>(`/agents/${agentId}/mcp-mounts`);
  },

  createAgentMount(agentId: string, payload: { server_id: string; approval_policy?: McpApprovalPolicy }) {
    return httpClient.post<McpAgentMount>(`/agents/${agentId}/mcp-mounts`, payload);
  },

  updateAgentMount(agentId: string, mountId: string, payload: { approval_policy?: McpApprovalPolicy; is_active?: boolean; apply_latest_snapshot?: boolean }) {
    return httpClient.patch<McpAgentMount>(`/agents/${agentId}/mcp-mounts/${mountId}`, payload);
  },

  deleteAgentMount(agentId: string, mountId: string) {
    return httpClient.delete<{ ok: boolean }>(`/agents/${agentId}/mcp-mounts/${mountId}`);
  },
};

File: backend/app/api/schemas/mcp.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class McpServerCreateRequest(BaseModel):
    name: str
    description: str | None = None
    server_url: str
    auth_mode: str = "none"
    static_bearer_token: str | None = None
    static_headers: dict[str, str] | None = None
    auth_config: dict[str, Any] = Field(default_factory=dict)
    oauth_client_id: str | None = None
    oauth_client_secret: str | None = None
    is_active: bool = True


class McpServerUpdateRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    server_url: str | None = None
    auth_mode: str | None = None
    static_bearer_token: str | None = None
    static_headers: dict[str, str] | None = None
    auth_config: dict[str, Any] | None = None
    oauth_client_id: str | None = None
    oauth_client_secret: str | None = None
    is_active: bool | None = None


class McpServerResponse(BaseModel):
    id: UUID
    organization_id: UUID
    name: str
    description: str | None
    server_url: str
    transport: str
    auth_mode: str
    is_active: bool
    auth_config: dict[str, Any]
    auth_metadata: dict[str, Any]
    capability_snapshot: dict[str, Any]
    oauth_client_id: str | None
    oauth_client_registration: dict[str, Any]
    tool_snapshot_version: int
    sync_status: str
    sync_error: str | None
    last_tested_at: datetime | None
    last_synced_at: datetime | None
    created_at: datetime
    updated_at: datetime
    has_static_bearer_token: bool
    has_static_headers: bool
    has_oauth_client_secret: bool


class McpDiscoveredToolResponse(BaseModel):
    id: UUID
    server_id: UUID
    snapshot_version: int
    name: str
    title: str | None
    description: str | None
    input_schema: dict[str, Any]
    annotations: dict[str, Any]
    tool_metadata: dict[str, Any]
    created_at: datetime


class McpAccountConnectionResponse(BaseModel):
    server_id: UUID
    user_id: UUID
    status: str
    scopes: list[Any]
    account_metadata: dict[str, Any]
    last_error: str | None
    access_token_expires_at: datetime | None
    refresh_token_expires_at: datetime | None
    last_refreshed_at: datetime | None
    last_used_at: datetime | None
    created_at: datetime
    updated_at: datetime


class McpAgentMountCreateRequest(BaseModel):
    server_id: UUID
    approval_policy: str = "ask"


class McpAgentMountUpdateRequest(BaseModel):
    approval_policy: str | None = None
    is_active: bool | None = None
    apply_latest_snapshot: bool = False


class McpAgentMountResponse(BaseModel):
    id: UUID
    agent_id: UUID
    server_id: UUID
    applied_snapshot_version: int
    approval_policy: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class McpAuthStartResponse(BaseModel):
    authorization_url: str
